[JSC] Make Date.UTC simpler

Constellation opened the pull request:
#### 445da2128acf942ea9ae00c179a31d33c590b934
<pre>
[JSC] Make Date.UTC simpler
<a href="https://bugs.webkit.org/show_bug.cgi?id=243446">https://bugs.webkit.org/show_bug.cgi?id=243446</a>

Reviewed by Ross Kirsling.

Avoid creating GregorianDateTime to generate ms from Date.UTC / Date constructor arguments
and making computation simpler and aligned to the spec.

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::millisecondsFromComponents):
(JSC::constructDate):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::localTimeToMS):
* Source/JavaScriptCore/runtime/JSDateMath.h:

Canonical link: <a href="https://commits.webkit.org/253044@main">https://commits.webkit.org/253044@main</a>
</pre>

rkirsling commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
     VM& vm = globalObject->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
 
+    auto toIntegerOrInfinity = [](double d) {
+        return trunc(std::isnan(d) ? 0.0 : d + 0.0);
+    };
```
If we already confirmed `isfinite`, can't this be inlined with the assumption that it's not `NaN`?

Constellation replied:
I wonder if we should keep it since `std::isnan` is super cheap anyway, and it makes `toIntegerOrInfinity` aligned to its implementation.

rkirsling commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
+        return trunc(std::isnan(d) ? 0.0 : d + 0.0);
+    };
+
+    // https://tc39.es/ecma262/#sec-maketime
+    auto makeTime = [](double hour, double min, double sec, double ms) {
+        return ((hour * msPerHour + min * msPerMinute) + sec * msPerSecond) + ms;
+    };
+
+    // https://tc39.es/ecma262/#sec-makeday
+    auto makeDay = [](double year, double month, double date) {
+        double additionalYears = std::floor(month / 12);
+        double ym = year + additionalYears;
+        if (!std::isfinite(ym))
+            return PNaN;
+        double mm = month - additionalYears * 12;
+        int32_t yearInt32 = toInt32(ym);
+        int32_t monthInt32 = toInt32(mm);
+        if (yearInt32 != ym || monthInt32 != mm)
+            return PNaN;
+        double days = dateToDaysFrom1970(yearInt32, monthInt32, 1);
+        return days + date - 1;
+    };
+
+    // https://tc39.es/ecma262/#sec-makedate
+    auto makeDate = [](double day, double time) {
+        if (!std::isfinite(day) || !std::isfinite(time))
+            return PNaN;
+        return day * msPerDay + time;
+    };
```
Aren't these operations used elsewhere in the Date API? How come we need to define them here?

Constellation replied:
The other places are using it too, but I'm currently only focusing on Date constructor and Date.UTC, and I also currently have no idea whether the other place should do it too, or not.

rkirsling reviewed (approved):
r=me with a couple of questions

Constellation committed:
[JSC] Make Date.UTC simpler
https://bugs.webkit.org/show_bug.cgi?id=243446

Reviewed by Ross Kirsling.

Avoid creating GregorianDateTime to generate ms from Date.UTC / Date constructor arguments
and making computation simpler and aligned to the spec.

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::millisecondsFromComponents):
(JSC::constructDate):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::localTimeToMS):
* Source/JavaScriptCore/runtime/JSDateMath.h:

Canonical link: https://commits.webkit.org/253044@main

darinadler commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
     auto scope = DECLARE_THROW_SCOPE(vm);
 
+    auto toIntegerOrInfinity = [](double d) {
+        return trunc(std::isnan(d) ? 0.0 : d + 0.0);
```
Seems a little strange to write `trunc` and `std::isnan`. How about `std::trunc`?

darinadler commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
+        int32_t monthInt32 = toInt32(mm);
+        if (yearInt32 != ym || monthInt32 != mm)
+            return PNaN;
+        double days = dateToDaysFrom1970(yearInt32, monthInt32, 1);
+        return days + date - 1;
```
Not sure this `days` local variable is doing us much good.

darinadler commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
+    // https://tc39.es/ecma262/#sec-makeday
+    auto makeDay = [](double year, double month, double date) {
+        double additionalYears = std::floor(month / 12);
+        double ym = year + additionalYears;
```
To me `ym` doesn’t say anything. Maybe `adjustedYear` or `yearIncludingMonths`. Unless `ym` is straight out the specification, in which case I am OK with this.

Constellation replied:
Yeah, this `ym` is exact name used for this value in the spec.

darinadler commented on Source/JavaScriptCore/runtime/DateConstructor.cpp:
```diff
@@ -71,6 +71,37 @@ static double millisecondsFromComponents(JSGlobalObject* globalObject, const Arg
+        double ym = year + additionalYears;
+        if (!std::isfinite(ym))
+            return PNaN;
+        double mm = month - additionalYears * 12;
```
`adjustedMonth` or `monthWithinYear`

Constellation replied:
Ditto. This is aligned to the spec linked in the comment.

darinadler reviewed (commented):
I guess you already landed this, but I reviewed it anyway.
